core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleReader.scala

```diff
@@ -33,17 +34,52 @@ private[spark] class HashShuffleReader[K, C](
     "Hash shuffle currently only supports fetching one partition")
 
   private val dep = handle.dependency
+  private val blockManager = SparkEnv.get.blockManager
 
   /** Read the combined key-values for this reduce task */
   override def read(): Iterator[Product2[K, C]] = {
+    val blockStreams = BlockStoreShuffleFetcher.fetchBlockStreams(
+      handle.shuffleId, startPartition, context)
+
+    // Wrap the streams for compression based on configuration
+    val wrappedStreams = blockStreams.map { case (blockId, inputStream) =>
+      blockManager.wrapForCompression(blockId, inputStream)
+    }
+
     val ser = Serializer.getSerializer(dep.serializer)
-    val iter = BlockStoreShuffleFetcher.fetch(handle.shuffleId, startPartition, context, ser)
+    val serializerInstance = ser.newInstance()
+
+    // Create a key/value iterator for each stream
+    val recordIterator = wrappedStreams.flatMap { wrappedStream =>
+      serializerInstance.deserializeStream(wrappedStream).asKeyValueIterator
+    }
+
+    val readMetrics = context.taskMetrics.createShuffleReadMetricsForDependency()
+    // Update read metrics for each record materialized
+    val metricIter = new InterruptibleIterator[(Any, Any)](context, recordIterator) {
+      override def next(): (Any, Any) = {
+        readMetrics.incRecordsRead(1)
+        delegate.next()
+      }
+    }
+
+    val iter = CompletionIterator[(Any, Any), Iterator[(Any, Any)]](metricIter, {
```

Comment:
This is a super-nit but if you change the above thing anyway, would you mind naming these iterators in a consistent way (so either all "iter" or all "iterator")?